File: backend/src/services/ShiftService.js
import { v4 as uuidv4 } from 'uuid';
import { query } from '../database/db.js';
import { ValidationError, NotFoundError } from '../utils/errors.js';

class ShiftService {
  /**
   * Open a new cashier shift
   */
  async openShift(cashierId, openingBalance = 0, notes = null) {
    // Check if cashier already has an active open shift
    const active = await query(
      'SELECT id FROM shifts WHERE cashier_id = ? AND status = "open"',
      [cashierId]
    );

    if (active.length > 0) {
      throw new ValidationError('Cashier already has an active open shift. Please close it first.');
    }

    const shiftId = uuidv4();
    await query(
      `INSERT INTO shifts (
        id, cashier_id, open_time, opening_balance, status, notes, created_at, updated_at
      ) VALUES (?, ?, NOW(), ?, 'open', ?, NOW(), NOW())`,
      [shiftId, cashierId, openingBalance, notes]
    );

    return this.getShiftById(shiftId);
  }

  /**
   * Get shift by ID
   */
  async getShiftById(shiftId) {
    const shifts = await query(
      `SELECT s.*, u.first_name as cashier_first_name, u.last_name as cashier_last_name, u.email as cashier_email
       FROM shifts s
       JOIN users u ON s.cashier_id = u.id
       WHERE s.id = ?`,
      [shiftId]
    );

    if (shifts.length === 0) {
      throw new NotFoundError('Shift');
    }

    return this.formatShift(shifts[0]);
  }

  /**
   * Get active shift for a cashier
   */
  async getActiveShift(cashierId) {
    const shifts = await query(
      `SELECT s.*, u.first_name as cashier_first_name, u.last_name as cashier_last_name, u.email as cashier_email
       FROM shifts s
       JOIN users u ON s.cashier_id = u.id
       WHERE s.cashier_id = ? AND s.status = 'open'
       ORDER BY s.open_time DESC
       LIMIT 1`,
      [cashierId]
    );

    if (shifts.length === 0) {
      return null;
    }

    const shift = shifts[0];

    // Compute live metrics for this active shift
    const stats = await query(
      `SELECT COUNT(*) as tx_count, 
              SUM(CASE WHEN method = 'cash' THEN amount ELSE 0 END) as cash_total,
              SUM(amount) as total_collected
       FROM payments
       WHERE issued_by = ? AND created_at >= ? AND status = 'paid'`,
      [cashierId, shift.open_time]
    );

    const liveStats = stats[0] || {};
    const expectedClosing = (shift.opening_balance || 0) + (liveStats.cash_total || 0);

    return {
      ...this.formatShift(shift),
      liveTransactionCount: liveStats.tx_count || 0,
      liveCashCollected: liveStats.cash_total || 0,
      liveTotalCollected: liveStats.total_collected || 0,
      liveExpectedCash: expectedClosing,
    };
  }

  /**
   * Close an active shift and reconcile drawer
   */
  async closeShift(shiftId, actualClosingBalance, notes = null) {
    const shifts = await query('SELECT * FROM shifts WHERE id = ?', [shiftId]);

    if (shifts.length === 0) {
      throw new NotFoundError('Shift');
    }

    const shift = shifts[0];
    if (shift.status === 'closed') {
      throw new ValidationError('Shift is already closed');
    }

    // Calculate cash payments during shift
    const stats = await query(
      `SELECT COUNT(*) as tx_count, 
              SUM(CASE WHEN method = 'cash' THEN amount ELSE 0 END) as cash_total
       FROM payments
       WHERE issued_by = ? AND created_at >= ? AND status = 'paid'`,
      [shift.cashier_id, shift.open_time]
    );

    const txCount = stats[0]?.tx_count || 0;
    const cashTotal = stats[0]?.cash_total || 0;
    const expectedBalance = (shift.opening_balance || 0) + cashTotal;
    const closingNotes = notes || shift.notes;

    await query(
      `UPDATE shifts SET 
        close_time = NOW(),
        expected_closing_balance = ?,
        actual_closing_balance = ?,
        transaction_count = ?,
        status = 'closed',
        notes = ?,
        updated_at = NOW()
       WHERE id = ?`,
      [expectedBalance, actualClosingBalance, txCount, closingNotes, shiftId]
    );

    const closed = await this.getShiftById(shiftId);
    const discrepancy = actualClosingBalance - expectedBalance;

    return {
      ...closed,
      discrepancy,
      discrepancyStatus: discrepancy === 0 ? 'balanced' : discrepancy > 0 ? 'overage' : 'shortage',
    };
  }

  /**
   * Get all shifts with filtering (Admin / POS Admin)
   */
  async getAllShifts(filters = {}, page = 1, limit = 20) {
    const { cashierId, status, startDate, endDate } = filters;
    let whereClause = 'WHERE 1=1';
    const params = [];

    if (cashierId) {
      whereClause += ' AND s.cashier_id = ?';
      params.push(cashierId);
    }

    if (status) {
      whereClause += ' AND s.status = ?';
      params.push(status);
    }

    if (startDate) {
      whereClause += ' AND DATE(s.open_time) >= ?';
      params.push(startDate);
    }

    if (endDate) {
      whereClause += ' AND DATE(s.open_time) <= ?';
      params.push(endDate);
    }

    const offset = (page - 1) * limit;

    const countResult = await query(
      `SELECT COUNT(*) as total FROM shifts s ${whereClause}`,
      params
    );
    const total = countResult[0]?.total || 0;

    const shifts = await query(
      `SELECT s.*, u.first_name as cashier_first_name, u.last_name as cashier_last_name, u.email as cashier_email
       FROM shifts s
       JOIN users u ON s.cashier_id = u.id
       ${whereClause}
       ORDER BY s.open_time DESC
       LIMIT ? OFFSET ?`,
      [...params, limit, offset]
    );

    return {
      data: shifts.map((s) => this.formatShift(s)),
      pagination: {
        page,
        limit,
        total,
        pages: Math.ceil(total / limit),
      },
    };
  }

  formatShift(shift) {
    const opening = parseFloat(shift.opening_balance || 0);
    const expected = shift.expected_closing_balance !== null ? parseFloat(shift.expected_closing_balance) : null;
    const actual = shift.actual_closing_balance !== null ? parseFloat(shift.actual_closing_balance) : null;
    const diff = expected !== null && actual !== null ? actual - expected : null;

    return {
      id: shift.id,
      cashierId: shift.cashier_id,
      cashierName: `${shift.cashier_first_name || ''} ${shift.cashier_last_name || ''}`.trim() || 'Cashier',
      cashierEmail: shift.cashier_email,
      openTime: shift.open_time,
      closeTime: shift.close_time,
      openingBalance: opening,
      expectedClosingBalance: expected,
      actualClosingBalance: actual,
      discrepancy: diff,
      transactionCount: shift.transaction_count || 0,
      status: shift.status,
      notes: shift.notes,
      createdAt: shift.created_at,
      updatedAt: shift.updated_at,
    };
  }
}

export default new ShiftService();

File: pos-system/POSSystem/Services/ShiftService.cs
using System;
using System.Collections.Generic;
using RMS.POSSystem.Database;
using RMS.POSSystem.Models;

namespace RMS.POSSystem.Services
{
    public class ShiftService
    {
        private readonly POSDatabase _db;
        public static CashierShift ActiveShift { get; private set; }

        public ShiftService(POSDatabase db)
        {
            _db = db;
        }

        public CashierShift GetActiveShift(string cashierId)
        {
            var rows = _db.ExecuteQuery(
                "SELECT * FROM Shifts WHERE CashierId = @cid AND Status = 'open' ORDER BY OpenTime DESC LIMIT 1",
                new Dictionary<string, object> { { "cid", cashierId } }
            );

            if (rows.Count == 0)
            {
                ActiveShift = null;
                return null;
            }

            var r = rows[0];
            ActiveShift = new CashierShift
            {
                Id = r["Id"]?.ToString(),
                CashierId = r["CashierId"]?.ToString(),
                OpenTime = r["OpenTime"] != null ? Convert.ToDateTime(r["OpenTime"]) : DateTime.UtcNow,
                OpeningBalance = r["OpeningBalance"] != null ? Convert.ToDecimal(r["OpeningBalance"]) : 0m,
                Status = "open",
                Notes = r["Notes"]?.ToString()
            };

            return ActiveShift;
        }

        public CashierShift OpenShift(string cashierId, decimal openingBalance, string notes)
        {
            var existing = GetActiveShift(cashierId);
            if (existing != null)
            {
                throw new InvalidOperationException("Cashier already has an active open shift.");
            }

            var shiftId = Guid.NewGuid().ToString();
            var now = DateTime.UtcNow;

            _db.ExecuteNonQuery(
                @"INSERT INTO Shifts (Id, CashierId, OpenTime, OpeningBalance, Status, Notes)
                  VALUES (@id, @cid, @open, @balance, 'open', @notes)",
                new Dictionary<string, object>
                {
                    { "id", shiftId },
                    { "cid", cashierId },
                    { "open", now },
                    { "balance", openingBalance },
                    { "notes", notes ?? "Shift started" }
                }
            );

            ActiveShift = new CashierShift
            {
                Id = shiftId,
                CashierId = cashierId,
                OpenTime = now,
                OpeningBalance = openingBalance,
                Status = "open",
                Notes = notes
            };

            return ActiveShift;
        }

        public CashierShift CloseShift(string shiftId, decimal actualCashCount, string closingNotes)
        {
            var rows = _db.ExecuteQuery("SELECT * FROM Shifts WHERE Id = @id", new Dictionary<string, object> { { "id", shiftId } });
            if (rows.Count == 0) throw new InvalidOperationException("Shift not found.");

            var shift = rows[0];
            var opening = Convert.ToDecimal(shift["OpeningBalance"]);
            var openTime = Convert.ToDateTime(shift["OpenTime"]);
            var cashierId = shift["CashierId"]?.ToString();

            // Calculate cash sales collected in this shift
            var txRows = _db.ExecuteQuery(
                @"SELECT COUNT(*) as Count, SUM(AmountPaid) as TotalCash 
                  FROM Transactions 
                  WHERE CashierId = @cid AND PaymentMethod = 'cash' AND CreatedAt >= @openTime",
                new Dictionary<string, object> { { "cid", cashierId }, { "openTime", openTime } }
            );

            int txCount = txRows.Count > 0 && txRows[0]["Count"] != null ? Convert.ToInt32(txRows[0]["Count"]) : 0;
            decimal totalCash = txRows.Count > 0 && txRows[0]["TotalCash"] != null ? Convert.ToDecimal(txRows[0]["TotalCash"]) : 0m;
            decimal expectedBalance = opening + totalCash;

            var closeTime = DateTime.UtcNow;

            _db.ExecuteNonQuery(
                @"UPDATE Shifts SET 
                    CloseTime = @closeTime,
                    ExpectedClosingBalance = @exp,
                    ActualClosingBalance = @act,
                    TransactionCount = @cnt,
                    Status = 'closed',
                    Notes = @notes
                  WHERE Id = @id",
                new Dictionary<string, object>
                {
                    { "closeTime", closeTime },
                    { "exp", expectedBalance },
                    { "act", actualCashCount },
                    { "cnt", txCount },
                    { "notes", closingNotes ?? "" },
                    { "id", shiftId }
                }
            );

            ActiveShift = null;

            return new CashierShift
            {
                Id = shiftId,
                CashierId = cashierId,
                OpenTime = openTime,
                CloseTime = closeTime,
                OpeningBalance = opening,
                ExpectedClosingBalance = expectedBalance,
                ActualClosingBalance = actualCashCount,
                TransactionCount = txCount,
                Status = "closed",
                Notes = closingNotes
            };
        }

        public List<CashierShift> GetAllShifts()
        {
            var list = new List<CashierShift>();
            var rows = _db.ExecuteQuery("SELECT * FROM Shifts ORDER BY OpenTime DESC LIMIT 100");

            foreach (var r in rows)
            {
                list.Add(new CashierShift
                {
                    Id = r["Id"]?.ToString(),
                    CashierId = r["CashierId"]?.ToString(),
                    OpenTime = Convert.ToDateTime(r["OpenTime"]),
                    CloseTime = r["CloseTime"] != DBNull.Value && r["CloseTime"] != null ? Convert.ToDateTime(r["CloseTime"]) : (DateTime?)null,
                    OpeningBalance = Convert.ToDecimal(r["OpeningBalance"]),
                    ExpectedClosingBalance = r["ExpectedClosingBalance"] != DBNull.Value && r["ExpectedClosingBalance"] != null ? Convert.ToDecimal(r["ExpectedClosingBalance"]) : 0m,
                    ActualClosingBalance = r["ActualClosingBalance"] != DBNull.Value && r["ActualClosingBalance"] != null ? Convert.ToDecimal(r["ActualClosingBalance"]) : 0m,
                    TransactionCount = r["TransactionCount"] != DBNull.Value && r["TransactionCount"] != null ? Convert.ToInt32(r["TransactionCount"]) : 0,
                    Status = r["Status"]?.ToString(),
                    Notes = r["Notes"]?.ToString()
                });
            }

            return list;
        }
    }
}
